portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -52,15 +60,27 @@ public void init() {
   }
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int maxComments = Integer.parseInt(request.getParameter("max_comments"));
-    Query comments_query = new Query("Comment").addSort("time_millis", SortDirection.DESCENDING);
-    List<Entity> results =  datastore.prepare(comments_query).asList(FetchOptions.Builder.withLimit(maxComments));
+    int commentsPerPage = Integer.parseInt(request.getParameter("comments_per_page"));
+    int pgNumber = Integer.parseInt(request.getParameter("pg_number")) - 1;
+    Query commentsQuery = new Query("Comment").addSort("time_millis", SortDirection.DESCENDING);
+    List<Entity> results = datastore
+      .prepare(commentsQuery)
+      .asList(FetchOptions.Builder
+      .withLimit(commentsPerPage)
+      .offset(commentsPerPage * pgNumber));
     List<Comment> comments = new ArrayList<Comment>();
     for(Entity comment : results){
       comments.add(makeComment(comment));
     }
     response.setContentType("application/json;");
-    response.getWriter().println(convertToJsonUsingGson(comments));
+    response.getWriter().println(convertToJsonUsingGson(new Response(comments, getLastPage(commentsPerPage))));
+  }
+  
+  // not very efficient, can improve with another entity to track total number of comments
+  private int getLastPage(double commentsPerPage){
+    Query commentsQuery = new Query("Comment");
+    List<Entity> results =  datastore.prepare(commentsQuery).asList(FetchOptions.Builder.withDefaults());
```

Comment:
This API, "countEntities()" is probably more idiomatic.

https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/PreparedQuery.html#countEntities-com.google.appengine.api.datastore.FetchOptions-

Reply:
yes, great find